### ISSUE TITLE
Discovery

### ISSUE THREAD

**Issue by garbled1** [jkeljo/greeneye-monitor#9]
This is not an issue with this library, but seemed the best way to contact you.

https://github.com/garbled1/py30303_disc

I just extended this library to do wiznet discovery, which means, theoretically, you could use it to discover greeneye's on the network for the (hopefully) future config_flow in HA.  Example:

```
[venv] [:~/git/py30303_disc/src] (master *%=)$ python -m py30303_disc.d30303_discover
INFO:py30303_disc.libs.py30303_disc:Listening on port 5001:udp for discovery events
DEBUG:py30303_disc.libs.py30303_disc:Initiating type wiznet discovery.
DEBUG:py30303_disc.libs.py30303_disc:Got Data: b'IMIN\x00\x08\xdc\x15\xff\xff\x01\xc0\xa8\n3\xff\xff\xff\x00\xc0\xa8\n\x0b\x00P\xc0\xa8\nK\x1f\x94\xff\x08\x00\x01\x00\x00\x07\x08\x00\n\x00\x00\x01\x04\x00\x00\x00\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x01@@@\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00'
DEBUG:py30303_disc.libs.py30303_disc:From ADDR: ('192.168.XX.51', 1460)
DEBUG:py30303_disc.libs.py30303_disc:Closing socket and ending discovery.
[{'ip_addr': '192.168.XX.51', 'mac_addr': '00-08-DC-XX-XX-XX', 'op_mode': 1, 'port': 8084}]
```

Now, it's not 100% perfect, because the wiznet device doesn't have a greeneye-unique macaddr, however, I'm returning op_mode, which, if set to 1, means the device is set in the mode you expect, which is the server mode.  But it gets you 90% of the way there.

Also.. and this is probably way more fiddly than you want to get in your own api library, but the wiznet protocol will allow you to actually configure the settings you require, rather than asking the user to do so. My library doesn't do that, but, you should know it's an option.  I've done it before in C.

**Comment by sdwilsh** [comment]
I think @jkeljo is using the support in siobrultech-protocols to configure the device right now (as I'm doing in brultech-serial2mqtt as well).

wiznet would certainly help with discovery in home assistant, however!

**Comment by garbled1** [comment]
Eons ago I had a C program that talked to the brultech gem.  In it, I used a wiznet reset command, to basically reboot the wiznet when it would stop responding.

Mine still hangs like that once in awhile.  It would certainly be nice, but not a must-have.  I can link to my code if such a thing interests you.

**Comment by jkeljo** [comment]
Interesting. My plan had been to do DHCP discovery, although like you say there does not appear to be a greeneye-unique MAC (they list a prefix in one of the setup docs, but it doesn't match the MAC that my device has). I'll probably still do that, but my bet is that it won't work for everybody. E.g. I wasn't aware of the wiznet option -- your library doesn't detect my Wifi + Ethernet GEM, and I'm guessing the DHCP discovery approach looking for a hostname starting with "greeneye" wouldn't find the wiznet ones. :(

For configuration, like @sdwilsh says I am planning to use the GEM API support that he and I built in siobrultech-protocols to do as much configuration as it will let me do. Unfortunately the GEM API doesn't including setting the IP and port of the server, though. 

The lovely folks at Brultech suggested that I could just manually POST to the relevant page of the web UI, and that does indeed work for mine, though this thread makes me think that, like the discovery, it likely won't work for everyone.

I've got [one more plumbing PR](https://github.com/home-assistant/core/pull/59871) waiting to be merged over at homeassistant before I can start putting up PRs for the actual UI config flow, discovery, etc. (If one of you guys want to go review that, it'll bump it back to the top of the queue and the core guys might get to it sooner.) The no-stacking rule they have over there is really frustrating :(; I could have had this done months ago if I didn't have to serialize.

**Comment by garbled1** [comment]
Yeah, the wiznet method does allow you to do that, but I'm guessing they don't actually USE a wiznet on the wifi one.. which is interesting, I would have thought they would use a wiznet for both....  Mine is probably older than yours, as I have a pretty low serial number, and wifi wasn't an option when I purchased it.

**Comment by sdwilsh** [comment]
I have an ethernet-only GEM, and I can't even send binary packets, as far as I can tell at least, like the ethernet/wifi one can.  Mine is fairly new too.

**Comment by garbled1** [comment]
https://github.com/garbled1/gnhast/tree/master/brulcoll

That's the C code I had for talking directly to the brultech.  In my case, I was using the client mode, rather than the server mode that you use here.  It's easier to setup, but I don't know if it offers any real benefit one way or the other.

That code also has the hacks in it for using the wiznet reboot.

**Comment by garbled1** [comment]
Absolutely.  It's older stuff, and I don't really work on it anymore.  I think it was one of the first open-source brultech integrations :)

**Comment by sarahmva** [comment]
The Wiznet reboot would be an amazing feature... I also suffer from the Wiznet hangs and I am looking for ways to reboot the device. Thinking about it, I could just use a zwave relay and power cycle it...
